puppet/ha-puppet/screenshot.js

```diff
@@ -238,9 +270,47 @@ export class Browser {
       }
 
       try {
-        // Wait for the page to be loaded.
+        // Wait for the page to load completely, to the extent it's possible to check..
         await page.waitForFunction(
-          () => {
+          async () => {
+            const deepFindIgnoreTags = ["svg", "slot", "style"];
+            const deepQuerySelectorAll = (root, selector, single = false) => {
+              const results = [];
+              function traverse(node) {
+                if (single && results.length > 0) {
+                  return;
+                }
+                if (node instanceof Element) {
+                  if (node.shadowRoot) {
+                    traverse(node.shadowRoot);
+                  }
+                  Array.from(node.children).forEach(child => {
+                    if (single && results.length > 0) return;
+                    if (child.matches(selector)) {
+                      results.push(child);
+                      if (single) return;
+                    }
+                    traverse(child);
+                  });
+                } else if (node instanceof DocumentFragment) {
+                  Array.from(node.children).forEach(child => {
+                    if (single && results.length > 0) return;
+                    if (child.matches(selector)) {
+                      results.push(child);
+                      if (single) return;
+                    }
+                    traverse(child);
+                  });
+                }
+              }
+              if (!deepFindIgnoreTags.includes(root.tagName.toLowerCase())) {
+                traverse(root);
+              }
+              return single ? (results[0] || null) : results;
+            }
+            const deepQuerySelector = (root, selector) => deepQuerySelectorAll(root, selector, true);
+
+
             const haEl = document.querySelector("home-assistant");
```

Comment:
For a future PR, we can speed up subsequent loads of this method by assigning resolved elements to a var stored on window. Especially if you want to query all cards. No need to query all cards again until all the cache of current queried cards is done loading.